### PR TITLE
Enable update in config between checkpoint and restore for session cache

### PR DESCRIPTION
**Pull request by anjumfatima90**
for #26845

### CODE REVIEW

**Comment by anjumfatima90** [comment]
#build

**Comment by anjumfatima90** [comment]
#build

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/ConfigRefresher.java [hunk 1]
```diff
@@ -82,12 +79,7 @@ public class ConfigRefresher {
         metatypeTracker.open();
 
         changesEndedHook = new ChangesEndedHook();
```
Comment (tjwatson) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/ConfigRefresher.java:81]: We don't need to keep a reference to this hook after it is added.  Could just do:

```java
        CheckpointPhase.getPhase().addMultiThreadedHook(Integer.MIN_VALUE, new ChangesEndedHook());
```

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java [hunk 2]
```diff
@@ -149,7 +149,7 @@ private void cacheInit() {
         // Attempt lazy initialization if necessary
         try {
             if (cacheStoreService.cacheManager == null)
-                cacheStoreService.activateLazily();
+                cacheStoreService.activateLazily();
```
Comment (tjwatson) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java]: It would be nice to avoid adding unnecessary whitespace.  Can make this change without doing another build IMO.

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java [hunk 3]
```diff
@@ -173,7 +173,7 @@ private void cacheInit() {
 
             if (trace && tc.isDebugEnabled())
                 tcInvoke(cacheStoreService.tcCacheManager, "getCache", metaCacheName, "String", "ArrayList");
-
+
```
Comment (tjwatson) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java]: Same whitespace comment.

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheStoreService.java [hunk 4]
```diff
@@ -541,7 +548,7 @@ public void setCompletedPassivation(boolean isInProcessOfStopping) {
         completedPassivation = isInProcessOfStopping;
     }
 
-    protected void setLibrary(Library library) {
+    protected void setLibrary(Library library) {
```
Comment (tjwatson) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheStoreService.java]: whitespace.

**Comment by anjumfatima90** [comment]
#build

**Comment by anjumfatima90** [comment]
#build

**Comment by anjumfatima90** [comment]
#build

**Review by dazavala — DISMISSED** [review]
Changes look good, and for the most part I understand them.  I'm not quite clear whether `CheckpointImpl.waitForConfig()` is now vestigial in light of adding  `notification.waitForCompletion()` to `ConfigRefresher.changesEndedHook.restore()`.  Will follow up with you.

**Comment by anjumfatima90** [comment]
#build

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java [hunk 5]
```diff
@@ -297,7 +294,6 @@ public void notificationCreated(RuntimeUpdateManager updateManager, RuntimeUpdat
         if (RuntimeUpdateNotification.CONFIG_UPDATES_DELIVERED.equals(notification.getName()) && jvmRestore.compareAndSet(true, false)) {
             debug(tc, () -> "Processing config on restore.");
             final CountDownLatch localLatch = new CountDownLatch(1);
```
Comment (dazavala) [inline dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java]: - There's no need for `localLatch`.  The latch is seeded with value `1` and is then used by `waitForConfig()` to block the checkpoint thread until the either the latch is decremented to `0`, times out, or the thread is interrupted,  You removed `waitForConfig()` -- i.e. you removed the wait on the latch -- so there's no need to call `localLatch.countDown()` in the `CompletionListener` methods. 
- I like that the `CompletionListener` methods provide trace that define the window of the restore config update.  If that's not a concern, then you could remove the `RuntimeUpdateListener` support from `CheckpointImpl`.

**Review by dazavala — CHANGES_REQUESTED** [review]
@anjumfatima90 I requested further changes to `CheckpointImpl` for commit 72f7eaec4d65c59036aee58ab09d4305fcdb15ed.  Thank you.

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java (hunk 5 above)
Comment (anjumfatima90) [reply dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java]: Removed, thanks!

**Comment by anjumfatima90** [comment]
#build

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java (hunk 5 above)
Comment (dazavala) [reply dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java]: Looks good.  Thank you @anjumfatima90.

**Review by dazavala — DISMISSED** [review]
Changes in recent commit 4ef6927387d70559eb738f2af26a16c303079b86 LGTM.

**Review by SmithaSubbarao — COMMENTED** [review]
My high level understanding of these changes is that a new class, Activator is being used to track updates (to the config) for storing/caching  data from users' applications and components during a checkpoint and restore (as opposed to updating notifications after a restore). I'm confused on how the checkpoint hook changes in this PR so I will need to follow up about this with @anjumfatima90 and @dazavala .

**Comment by anjumfatima90** [comment]
#build

**Review by tjwatson — APPROVED** [review]

**Comment by tjwatson** [comment]
#libby

**Comment by anjumfatima90** [comment]
#build

**Comment by uitesterhero** [comment]
This pr has caused intermittent but often failures in two tests and will need to be reverted - see defect 299821
